beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,1012 @@
+		log.WithError(err).Error("Error encountered while walking data column filesystem.")
+	}
+
+	// Prune the cache and the filesystem.
+	dcs.prune()
+
+	log.WithField("elapsed", time.Since(start)).Info("Data column filesystem cache warm-up complete")
+}
+
+// Summary returns the DataColumnStorageSummary.
+func (dcs *DataColumnStorage) Summary(root [fieldparams.RootLength]byte) DataColumnStorageSummary {
+	return dcs.cache.Summary(root)
+}
+
+// Save saves data column sidecars into the database and asynchronously performs pruning.
+// The returned chanel is closed when the pruning is complete.
+func (dcs *DataColumnStorage) Save(dataColumnSidecars []blocks.VerifiedRODataColumn) error {
+	startTime := time.Now()
+
+	if len(dataColumnSidecars) == 0 {
+		return nil
+	}
+
+	// Check the number of columns is the one expected.
+	// While implementing this, we expect the number of columns won't change.
+	// If it does, we will need to create a new version of the data column sidecar file.
+	if params.BeaconConfig().NumberOfColumns != mandatoryNumberOfColumns {
+		return errWrongNumberOfColumns
+	}
+
+	highestEpoch := primitives.Epoch(0)
+	dataColumnSidecarsbyRoot := make(map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn)
+
+	// Group data column sidecars by root.
+	for _, dataColumnSidecar := range dataColumnSidecars {
+		// Check if the data column index is too large.
+		if dataColumnSidecar.Index >= mandatoryNumberOfColumns {
+			return errDataColumnIndexTooLarge
+		}
+
+		// Group data column sidecars by root.
+		root := dataColumnSidecar.BlockRoot()
+		dataColumnSidecarsbyRoot[root] = append(dataColumnSidecarsbyRoot[root], dataColumnSidecar)
+	}
+
+	for root, dataColumnSidecars := range dataColumnSidecarsbyRoot {
+		// Safety check all data column sidecars for this root are from the same slot.
+		firstSlot := dataColumnSidecars[0].SignedBlockHeader.Header.Slot
+		for _, dataColumnSidecar := range dataColumnSidecars[1:] {
+			if dataColumnSidecar.SignedBlockHeader.Header.Slot != firstSlot {
+				return errDataColumnSidecarsFromDifferentSlots
+			}
+		}
+
+		// Set the highest epoch.
+		epoch := slots.ToEpoch(dataColumnSidecars[0].Slot())
+		highestEpoch = max(highestEpoch, epoch)
+
+		// Save data columns in the filesystem.
+		if err := dcs.saveFilesystem(root, epoch, dataColumnSidecars); err != nil {
+			return errors.Wrap(err, "save filesystem")
+		}
+
+		// Get all indices.
+		indices := make([]uint64, 0, len(dataColumnSidecars))
+		for _, dataColumnSidecar := range dataColumnSidecars {
+			indices = append(indices, dataColumnSidecar.Index)
+		}
+
+		// Compute the data columns ident.
+		dataColumnsIdent := DataColumnsIdent{Root: root, Epoch: slots.ToEpoch(dataColumnSidecars[0].Slot()), Indices: indices}
+
+		// Set data columns in the cache.
+		if err := dcs.cache.set(dataColumnsIdent); err != nil {
+			return errors.Wrap(err, "cache set")
+		}
+
+		// Notify the data column feed.
+		dcs.DataColumnFeed.Send(dataColumnsIdent)
```

Comment:
This doesn't feel safe to me. Due to how the event feed is implemented, it's easy to get blocked by a slow reader on the other side, so I don't think it's a good idea to do this in the Save path. It would be safer to call the method in a goroutine, but I'm curious if there's another pattern that could work here. How do you plan for the event feed to be used?

Reply:
You mean `dcs.DataColumnFeed.Send` blocks until **all** subscribers retrieved the value?

Reply:
Goroutine added in 5be3a72a542f062d98d13c72f9e3240514093d36.

Reply:
I'll look more at how this is used on the other end to see if there's another tool besides event feed. As to reducing goroutines, you could do the following:
- on service init, create a single buffered channel
- on service init, spawn a single goroutine that reads from this new channel and writes to the event feed
- Save method can just write to this buffered channel, and we don't have a problem as long as the readers roughly keep up within the margins of the buffer

I think writing to a channel should also be faster than spawning a goroutine.

Reply:
I think I have a better solution involving no goroutine and no extra channel in 55a0de6747653257196f32a41b2687d3d569303d.

==> We force consumers to use a single buffered channel.